### ISSUE TITLE
Rugrats: Adventures in Gameland

### ISSUE THREAD

**Issue by kjetil-f** [farmerbb/RED-Project#178]
This game uses a NES ROM and adds hi-res graphics on top (similar to the HD packs for Mesen emulator). It exist a physical NES version of this game, and the Steam demo have a NES ROM of the demo included in the installation folder,

Is it possible to extract the full NES ROM from this game?

**Comment by RealRelativeEase** [comment]
Don't have the game myself, but according to [SteamDB](https://steamdb.info/depot/2633951/), there's an NES ROM (512 kilobytes) among the installation files.

**Comment by mjBrickman** [comment]
https://steamdb.info/depot/2633951/history/
According to SteamDB's history feature, the NES ROM was removed in Manifest 3495750955999334612.

**Comment by kjetil-f** [comment]
As far as I know, the NES ROM is only a demo anyways. However, since they seem to be using some kind of emulation on this release, I'm guessing the full ROM should be inside one of the files somewhere.

**Comment by farmerbb** [comment]
Yep, the ROM included with the demo on Steam only has the first two levels

**Comment by farmerbb** [comment]
Hashes of the ROM were posted at https://www.romhacking.net/forum/index.php?topic=38363.msg461710#msg461710 :

>CRC32: 97130010
>MD5: 90e1a66526fc624ca071c28a78b85b7a
>SHA1: 1345d8c4e5cc289630d7f0a517adf74e0237e79c

**Comment by farmerbb** [comment]
Looks like some people have been able to extract audioless versions of the ROM from the retail game:
https://steamcommunity.com/app/2633950/discussions/0/4757578733053996569/

**Comment by RealRelativeEase** [comment]
According to a more recent post on the same thread, the updated version still contains an extractable NES ROM with working audio: [https://steamcommunity.com/app/2633950/discussions/0/4757578733053996569/#c6655846466487575461](https://steamcommunity.com/app/2633950/discussions/0/4757578733053996569/#c6655846466487575461)

**Comment by bmsuseluda** [comment]
> According to a more recent post on the same thread, the updated version still contains an extractable NES ROM with working audio: https://steamcommunity.com/app/2633950/discussions/0/4757578733053996569/#c6655846466487575461

i can confirm this

**Comment by kjetil-f** [comment]
> i can confirm this
The full version or the demo?

**Comment by mjBrickman** [comment]
The Steam thread appears to have been taken down. The "steamworks_rugrats" 7z file also does not appear to be present in the Epic Games store release.

**Comment by bmsuseluda** [comment]
> > i can confirm this
> > The full version or the demo?

the full version

**Comment by kjetil-f** [comment]
Quote from another Steam thread: "Hmm, not sure why that topic got nuked. So, the "steamworks_rugrats" file is compressed with the .7z format. You can use 7-Zip to extract it. Once you get the uncompressed file (about 524 KB in size), add ".nes" to the end and it will load in an NES emulator as normal. It works like a normal NES rom, with full audio and no apparent issues thus far. Hope this helps!"

**Comment by kjetil-f** [comment]
Another quote on Steam: "if anyone comes into this topic looking for the rom, they changed it so you have to beat the game first then it appears in the install directory."

**Comment by Fabs-tinkers** [comment]
I used assetripper on Rugrats_Data/data.unity3d  and exported primari content to a new folder.  Rugrats nes-rom is in Textasset folder under name: steamworks_rugrats.nes.bytes - just rename it .nes and run it.

**Comment by PW5190** [comment]
I claimed it for free from the Epic Games Store. Is there a way to extract the NES ROM from this version?

**Comment by kjetil-f** [comment]
> I claimed it for free from the Epic Games Store. Is there a way to extract the NES ROM from this version?

It doesn't seem to be included in that version. I could not find it when using the assetripper tool at least.

**Comment by Fabs-tinkers** [comment]
> > I claimed it for free from the Epic Games Store. Is there a way to extract the NES ROM from this version?
> 
> It doesn't seem to be included in that version. I could not find it when using the assetripper tool at least.

Sadly I found no roms there - probably encrypted differently there. - Only steam version.

**Comment by TheNametag** [comment]
It would appear that they have altered the file and location a bit.  Now, after extracting the data.unity3d file, I had to go into sharedassets1.assets to find the NES file.  It is listed as a text asset still, "steamworks_rugrats.nes".  However, just renaming it to .nes didn't run, as there's 32 bits at the beginning, before the NES header, that I had to remove.  That at least loads it now, but I have not tested the full ROM this way to see if it works entirely.

**Comment by TheNametag** [comment]
The above was for the "Congratulations" version of the game, the one you get as a reward for beating it.

There are three other versions of the ROM inside sharedassets1.assets.  They are marked "rugrats" "rugrats-ps" and "rugrats-ps_Asia".  IDs 6302, 6301, and 6299 respectively (the Congratulations is 541).  Each one will need to have a bit of data before the "NES" removed in a hex editor, which varies in length depending on the file, around a few dozen bytes.  None of those have audio.  The differences are the original uses the Xbox key (a), and has the selection off of 8-bit (which does nothing), while the PS ones use the PS buttons, X for regular, O for Asia.  None of them contain any audio, meaning the "Congratulations" version above is the one you likely want to rip... and as mentioned you DO receive it in the base directory under "congrats_rugrats.nes" when you beat the game, which takes ~60-90 minutes your first time.

**Comment by farmerbb** [comment]
Thanks @TheNametag. I got the ROMs extracted from the Epic Games version (so no sound for me 🙃).  I am not an expert at Unity asset extraction, but here are the exact steps I followed in case it helps anyone else:

1. Open AssetStudioGUI, go to File -> Extract folder -> select the Rugrats_Data folder in your game install, then select a folder to extract the game data to.
2. When the extraction finishes, open Asset Bundle Extractor, go to File -> Open and select `sharedassets1.assets` in the `data.unity3d_unpacked` folder created in the previous step.
3. If a "Select a type database" dialog box pops up, just click OK.  (My version of UABE only supports up to `2021.3.0f1` and Rugrats is using `2022.3.28f1`.
4. Click the entry for sharedassets1.assets under "Files and Components", then sort the output by "Size (bytes)". The ROMs will be the three files at the top with a size of just over 524288 bytes
5. Shift-select these files and click "Export Raw", then select a directory to export the files to.
6. Open each extracted file in a hex editor, delete all bytes from the beginning of the file before the `NES` header, then save the file and rename it with a .nes file extension.
